models/Group.ts

```diff
@@ -126,6 +142,165 @@ export default function (sequelize, DataTypes): GroupStatic {
+        };
+      } else {
+        return {
+          name: (s as Station).name,
+          location: (s as Station).location.coordinates
+        }
+      }
+    });
+    for (const a of allStations) {
+      for (const b of allStations) {
+        if (a !== b && a.name !== b.name) {
+          if (latLngApproxDistance(a.location, b.location) < MAX_STATION_DISTANCE_THRESHOLD_METERS) {
+            throw new ClientError("Stations too close together");
+          }
+        }
+      }
+    }
+
+    // Add new stations, or update lat/lng if station with same name but different lat lng.
+    const addedStations = [];
+    for (const [name, station] of Object.entries(newStationsByName)) {
+      if (!existingStationsByName.hasOwnProperty(name)) {
+        const stationObj = new models.Station({
+          name: station.name,
+          location: [station.lat, station.lng],
+          lastUpdatedById: authUser.id
+        });
+        addedStations.push(stationObj);
+        stationOpsPromises.push(
+            new Promise(async (resolve) => {
+              await stationObj.save();
+              await group.addStation(stationObj);
+              resolve();
+            })
+        );
+      } else {
+        // Update lat/lng if it has changed but the name is the same
+        const stationToUpdate = existingStationsByName[station.name];
+
+        const EPSILON = 0.000000000001;
+        // We need to compare these numbers with an epsilon value, otherwise we get floating-point precision issues.
+        if (
+            Math.abs(stationToUpdate.location.coordinates[0] - station.lat) < EPSILON ||
+            Math.abs(stationToUpdate.location.coordinates[1] - station.lat) < EPSILON
+        ) {
+          // NOTE - Casting this as "any" because station.location has a special setter function
+          (stationToUpdate as any).location = [station.lat, station.lng];
+          stationToUpdate.lastUpdatedById = authUser.id;
+          addedStations.push(stationToUpdate);
+          stationOpsPromises.push(stationToUpdate.save());
+        } else if (applyToRecordingsFromDate) {
+          addedStations.push(stationToUpdate);
+        }
+      }
+    }
+
+    if (!applyToRecordingsFromDate) {
+      return await Promise.all([...stationOpsPromises, ...retirePromises]).then(() => addedStations.map(({id}) => id));
+    } else {
+      // After adding stations, we need to apply any station matches to recordings from a start date:
+
+      await Promise.all([...stationOpsPromises, ...retirePromises]);
+      // Now addedStations are properly resolved with ids:
+      // Now we can look for all recordings in the group back to startDate, and check if any of them
+      // should be assigned to any of our stations.
+
+      // Find matching recordings to apply stations to from `applyToRecordingsFromDate`
+
+      // Get recordings for group starting at date:
+      const builder = await new models.Recording.queryBuilder().init(
```

Comment:
Surely updateExistingRecordings() is another function.   I believe it should work in exactly the same way as when a recording comes in.  So instead of using the calculated version of addedStations (including all stations in this case) it should just go to the stations table - get all the stations and go from there.